In [3]:
import pandas as pd
import os
from sklearn.model_selection import train_test_split
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# Rocinante 36 Model

In [4]:
file_path = 'Exhibit+2.xlsx'
roc = pd.read_excel(file_path, sheet_name='Rocinante Models')
roc

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0    Rocinante 1                   171.877                      6.1   
1    Rocinante 2                   139.796                      6.1   
2    Rocinante 3                   178.947                      9.9   
3    Rocinante 4                   140.022                      5.8   
4    Rocinante 5                   186.476                     10.0   
5    Rocinante 6                   192.123                      6.5   
6    Rocinante 7                   175.085                      5.5   
7    Rocinante 8                   146.882                      8.4   
8    Rocinante 9                   202.847                      6.6   
9   Rocinante 10                   149.933                      8.8   
10  Rocinante 11                   171.579                      9.5   
11  Rocinante 12                   201.512                      9.2   
12  Rocinante 13                   142.604                      8.1   
13  Rocinante 14                   135.158                      5.4   
14  Rocinante 15                   107.322                      5.6   
15  Rocinante 16                   199.699                      6.3   
16  Rocinante 17                   202.875                      5.8   
17  Rocinante 18                   204.406                      7.7   
18  Rocinante 19                   107.331                      7.2   
19  Rocinante 20                   190.345                      8.4   
20  Rocinante 21                   108.419                      8.4   
21  Rocinante 22                   150.249                      5.7   
22  Rocinante 23                   155.882                      5.5   
23  Rocinante 24                   100.980                      8.4   
24  Rocinante 25                   182.679                      5.6   
25  Rocinante 26                   166.752                      9.9   
26  Rocinante 27                   121.561                      6.2   
27  Rocinante 28                   174.256                      9.9   
28  Rocinante 29                   119.018                      6.5   
29  Rocinante 30                   169.842                      7.1   
30  Rocinante 31                   198.311                      9.6   
31  Rocinante 32                   204.875                      5.8   
32  Rocinante 33                   119.561                      6.2   
33  Rocinante 34                   203.875                      5.9   
34  Rocinante 35                   118.561                      6.1   

    Mileage\n(Km/ltr)  Top speed (Km/hr)  
0                15.8              168.2  
1                12.1              149.6  
2                17.0              173.4  
3                11.6              170.6  
4                17.2              175.0  
5                17.6              173.1  
6                16.0              184.6  
7                13.0              175.7  
8                19.3              166.7  
9                13.3              175.4  
10               15.8              150.9  
11               19.2              140.5  
12               11.7              150.2  
13               11.2              146.2  
14                7.8              151.2  
15               19.0              143.2  
16               19.7              178.9  
17               19.5              168.2  
18                7.5              164.7  
19               18.1              182.7  
20                7.9              165.7  
21               12.6              137.0  
22               12.9              173.9  
23                7.6              165.1  
24               16.7              160.1  
25               15.5              155.6  
26                9.5              155.4  
27               16.3              161.9  
28                9.1              146.8  
29               15.3              149.5  
30               19.1              146.5  
31               19.4              178.9  
32                9.4              1

In [6]:
X_rocinante = roc[['Price\n(in lakh rupees)', 'Mileage\n(Km/ltr)', 'Top speed (Km/hr)']]
y_rocinante = roc['Sales \n(in 1,000 units)']

### Step-1: Split the data into training and testing sets (70% training, 30% testing)

In [7]:
X_train_roc, X_test_roc, y_train_roc, y_test_roc = train_test_split(X_rocinante, y_rocinante, test_size=0.3, random_state=42)

# Initialize and train a Linear Regression model
model_rocinante = LinearRegression()
model_rocinante.fit(X_train_roc, y_train_roc)

# Make predictions on the test set
y_pred_roc = model_rocinante.predict(X_test_roc)

### Step-2: Calculation of RMSE ( Root Mean Squared Error)

In [9]:
RMSE_roc = mean_squared_error(y_test_roc, y_pred_roc, squared=False)

# Print the RMSE result
print(f'Root Mean Squared Error(RMSE) for Rocinante Model : {RMSE_roc}')

Root Mean Squared Error(RMSE) for Rocinante Model : 2.284066588336076


#### The Root Mean Squared Error (RMSE) for the Rocinante Model is 2.28. This means that, on average, the model's predictions for Rocinante vehicle sales are off by about 2.28 units (in thousands) compared to the actual sales.

#### The model's typical prediction error is around 2,280 sales units. An RMSE of 2.28 suggests that the model is reasonably accurate, but there might still be some room for improvement to make the predictions closer to the actual sales figures.

### Predicting both training and testing data

In [10]:
y_train_pred = model_rocinante.predict(X_train_roc)
y_test_pred = model_rocinante.predict(X_test_roc)

### Calculation of RMSE for both training and testing data

In [11]:
train_rmse_roc = mean_squared_error(y_train_roc, y_train_pred)
test_rmse_roc = mean_squared_error(y_test_roc, y_test_pred)

In [12]:
print("Training RMSE:", train_rmse_roc)
print("Testing RMSE:", test_rmse_roc)

Training RMSE: 4.990464394910194
Testing RMSE: 5.216960179953201


### Step-3: Determining whether the model is Overfitting or Underfitting

In [13]:
if train_rmse_roc < test_rmse_roc:
    print("The model may be overfitting.")
elif train_rmse_roc > test_rmse_roc:
    print("The model may be underfitting.")
else:
    print("The model is performing consistently on both training and testing data.")

The model may be overfitting.


### The Training RMSE (4.99) is lower than the Testing RMSE (5.22).
#### The difference between the two values is relatively small, indicating that the model is not dramatically overfitting or underfitting.
### Interpretation:
#### Slight Overfitting: The model may be experiencing slight overfitting, as it performs better on the training data than on the testing data. This suggests it has learned the training data well but does not generalize perfectly to new, unseen data.

# Marengo32 Model

In [14]:
mar = pd.read_excel(file_path, sheet_name='Marengo Models')
mar

Cars  Sales \n(in 1,000 units)  Price\n(in lakh rupees)  \
0    Marengo 1                    20.896                     42.5   
1    Marengo 2                    31.048                     36.0   
2    Marengo 3                    29.904                     54.7   
3    Marengo 4                    28.792                     42.7   
4    Marengo 5                    16.776                     44.9   
5    Marengo 6                    18.928                     35.5   
6    Marengo 7                    22.776                     51.3   
7    Marengo 8                    36.824                     30.4   
8    Marengo 9                    22.216                     38.4   
9   Marengo 10                    35.456                     32.2   
10  Marengo 11                    28.576                     59.4   
11  Marengo 12                    33.648                     50.8   
12  Marengo 13                    33.440                     57.5   
13  Marengo 14                    34.912                     44.6   
14  Marengo 15                    24.016                     34.6   
15  Marengo 16                    20.992                     56.4   
16  Marengo 17                    21.696                     41.4   
17  Marengo 18                    17.832                     53.7   
18  Marengo 19                    25.624                     44.8   
19  Marengo 20                    21.920                     32.4   
20  Marengo 21                    24.792                     40.9   
21  Marengo 22                    22.256                     44.1   
22  Marengo 23                    15.488                     46.6   
23  Marengo 24                    21.640                     31.4   
24  Marengo 25                    23.264                     54.7   
25  Marengo 26                    19.328                     37.9   
26  Marengo 27                    27.216                     52.9   
27  Marengo 28                    23.096                     42.7   
28  Marengo 29                    17.824                     54.9   
29  Marengo 30                    19.144                     44.0   
30  Marengo 31                    21.600                     42.2   

    Mileage\n(Km/ltr)  Top speed (Km/hr)  
0                 9.3              199.4  
1                 9.7              235.2  
2                16.6              240.8  
3                11.7              232.5  
4                13.7              188.8  
5                 9.6              184.2  
6                13.7              207.7  
7                12.6              249.5  
8                16.2              175.8  
9                 9.6              245.6  
10               11.3              223.3  
11               13.6              240.0  
12               13.5              250.0  
13               11.1              247.4  
14               14.1              192.0  
15               13.4              201.2  
16               12.8              193.3  
17               16.0              202.0  
18               12.7              208.9  
19               13.6              176.9  
20               11.6              218.2  
21               11.4              190.0  
22               16.3              176.8  
23               14.5              178.9  
24               14.3              192.3  
25                9.4              184.1  
26               15.3              240.3  
27               15.0              192.1  
28               13.5              199.0  
29               11.8              181.1  
30                9.3              181.2

In [15]:
X_marengo  = mar[['Price\n(in lakh rupees)', 'Mileage\n(Km/ltr)', 'Top speed (Km/hr)']]
y_marengo  = mar['Sales \n(in 1,000 units)']

### Step-1: Split the data into training and testing sets (70% training, 30% testing)

In [17]:
X_train_mar, X_test_mar, y_train_mar, y_test_mar = train_test_split(X_marengo, y_marengo, test_size=0.3, random_state=42)

# Initialize and train a Linear Regression model
model_marengo = LinearRegression()
model_marengo.fit(X_train_mar, y_train_mar)

# Make predictions on the test set
y_pred_mar = model_marengo.predict(X_test_mar)

### Step-2: Calculation of RMSE ( Root Mean Squared Error)

In [18]:
RMSE_mar = mean_squared_error(y_test_mar, y_pred_mar, squared=False)
print(f'Root Mean Squared Error(RMSE) for Marengo Model : {RMSE_mar}')

Root Mean Squared Error(RMSE) for Marengo Model : 2.7322921310468495


#### The Root Mean Squared Error (RMSE) of 2.73 indicates that, on average, the predictions made by the Marengo Model deviate from the actual values by about 2.73 units. A lower RMSE generally reflects better predictive accuracy, but the acceptable level depends on the context and scale of the data.

### Predicting both training and testing data

In [19]:
y_train_pred = model_marengo.predict(X_train_mar)
y_test_pred = model_marengo.predict(X_test_mar)

### Calculation of RMSE for both training and testing data

In [20]:
train_rmse_mar = mean_squared_error(y_train_mar, y_train_pred)
test_rmse_mar = mean_squared_error(y_test_mar, y_test_pred)

In [21]:
print("Training RMSE:", train_rmse_mar)
print("Testing RMSE:", test_rmse_mar)

Training RMSE: 3.9768018689111426
Testing RMSE: 7.465420289380534


### Step-3: Determining whether the model is Overfitting or Underfitting

In [22]:
if train_rmse_mar < test_rmse_mar:
    print("The model may be overfitting.")
elif train_rmse_mar > test_rmse_mar:
    print("The model may be underfitting.")
else:
    print("The model is performing consistently on both training and testing data.")

The model may be overfitting.


### Interpretation:
#### The model shows a significantly lower RMSE on the training set (3.98) compared to the testing set (7.47). This indicates overfitting, where the model has learned the training data very well but fails to generalize to new, unseen data. The large gap between training and testing RMSE suggests that the model is too complex, capturing noise or patterns specific to the training set rather than underlying trends. Reducing model complexity or using regularization techniques could help mitigate overfitting.